pkg/tests/run.go

```diff
@@ -47,129 +49,124 @@ func (r *Runner) Run(ctx context.Context) error {
-			case "exponential":
-				ramper = ramp.NewRampingService(ramp.ExponentialRamp, currentStartRate, currentEndRate, currentSteps)
-			}
-		}
-
-		if ramper == nil {
-			r.logger.Info(ctx, "Executing Test: %s", t.TestName)
-			r.logger.Info(ctx, "Rate: %s", t.Rate.String())
-			r.logger.Info(ctx, "Duration: %s", t.Duration.String())
-			r.logger.Info(ctx, "Endpoint: %s", t.Path)
-			err = t.Handler(ctx, &t)
-			if err != nil {
-				return err
-			}
-		} else {
-			r.logger.Info(ctx, "Executing Test: %s", t.TestName)
-			r.logger.Info(ctx, "Ramp type: %s", ramper.GetType())
-			r.logger.Info(ctx, "Endpoint: %s", t.Path)
-			if currentRampDuration == 0 {
-				duration := math.Round(t.Duration.Minutes() / float64(ramper.GetSteps()))
-				t.Duration = time.Duration(duration) * time.Minute
-			} else {
-				remainingDuration = int(t.Duration.Minutes()) - currentRampDuration
-				duration := math.Round(float64(currentRampDuration) / float64(ramper.GetSteps()))
-				t.Duration = time.Duration(duration) * time.Minute
-			}
-
-			for i := 0; i < ramper.GetSteps(); i++ {
-				r.logger.Info(ctx, "Ramping up... step %v", i+1)
-				rateInt := ramper.NextRate()
-				newRate, _ := helpers.ParseRate(fmt.Sprint(rateInt))
-				t.Rate = newRate
-				if i+1 == ramper.GetSteps() && remainingDuration > 0 {
-					t.Duration = t.Duration + (time.Duration(remainingDuration) * time.Minute)
-				}
-				r.logger.Info(ctx, "Rate: %s", t.Rate.String())
-				r.logger.Info(ctx, "Duration: %s", t.Duration.String())
-				err = t.Handler(ctx, &t)
+		for i, conn := range r.connections {
+			go func(ctx context.Context, concurrentConnections int, index int, conn *sdk.Connection, testOptions types.TestOptions) error {
+				// Create an Attacker for each individual test. This is due to the
+				// fact that vegeta (and compatible parsers, such as benchmark-wrapper)
+				// expect the sequence to start at 0 for each result file. (Possibly a bug?)
+				connAttacker := vegeta.Client(&http.Client{Transport: conn})
+				attacker := vegeta.NewAttacker(connAttacker)
+
+				// Open a file and create an encoder that will be used to store the
+				// results for each test.
+				fileName := fmt.Sprintf("%s_%s_%d.json", r.testID, testOptions.TestName, index)
+				resultsFile, err := helpers.CreateFile(fileName, r.outputDirectory)
 				if err != nil {
 					return err
 				}
-			}
-		}
+				encoder := vegeta.NewJSONEncoder(resultsFile)
+
+				// Bind "Test Harness"
+				testOptions.ID = r.testID
+				testOptions.Attacker = attacker
+				testOptions.Connection = conn
+				testOptions.Encoder = &encoder
+				testOptions.Logger = r.logger
+
+				// Create the vegeta rate with the config values
+				currentTestRate := confHelper.ResolveStringConfig(ctx, rate, fmt.Sprintf("%s.rate", testOptions.TestName))
+				rate, err := helpers.ParseRate(currentTestRate, concurrentConnections)
+				if err != nil {
+					r.logger.Warn(ctx,
+						"error parsing rate for test %s: %s. Using default",
```

Comment:
should wee actually be failing here instead of just warning? when would we hit this and what are the potential traps?

Reply:
We would hit this when the configuration has a specific test rate, that should override the global rate.
If that is in wrong format (IMO that should be difficult to do) the parse will fail, but since there is a global rate, that has been already validated, instead of breaking the test execution we use the global rate.